test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs

```diff
@@ -2118,12 +2115,12 @@ public Task ResponseStatusCodeSetBeforeHttpContextDisposedRequestMalformedReadIg
                     // Don't wait to receive the response. Just close the socket.
                 }
 
-                connectionCloseWh.Set();
+                connectionCloseWh.TrySetResult(null);
 
                 // Write failed
-                await Assert.ThrowsAsync<TaskCanceledException>(async () => await writeTcs.Task);
+                await Assert.ThrowsAsync<TaskCanceledException>(async () => await writeTcs.Task).TimeoutAfter(TimeSpan.FromSeconds(15));
```

Comment:
I decided to switch back to original code (non tcs version) and just add another ManualResetEventSlim to fix the race condition.